File: Chapter 6 - Lecture/p10.m
%P10  Polynomials and corresponding equipotential curves
close all;
scrsz = get(groot,'ScreenSize');
figure('position', [150   100   0.8*scrsz(3:4)]), clf;

%%
N = 16; 
fs = 14;
for i = 1:2
    if i == 1, s = 'equispaced points'; x = -1 + 2*(0:N)/N; end
    if i == 2, s = 'Chebyshev points'; x = cos(pi*(0:N)/N); end
    p = poly(x);
    % Plot p(x) over [-1,1]:
    
    y = -1: 0.005: 1;
    q = polyval(p, y);
    subplot(2,2,2*i-1)
    plot(x, 0*x, 'o', 'MarkerFaceColor', 0.5*ones(1,3)), hold on
    plot(y, q, 'k-', 'linewidth', 0.8), grid on
    set(gca, 'xtick', -1:0.5:1)
    title(s, 'FontSize', fs)
    
    % Plot equipotential curves:
    subplot(2,2,2*i)
    plot(real(x), imag(x), 'o', 'MarkerFaceColor', 0.5*ones(1,3)), hold on
    axis([-1.4 1.4 -1.12 1.12])
    xgrid = linspace(-1.4, 1.4, 101);
    ygrid = linspace(-1.12, 1.12, 101);
    [xx, yy] = meshgrid(xgrid, ygrid);
    zz = xx + 1i*yy;
    pp = polyval(p, zz); levels = 10.^(-4:0);
    contour(xx, yy, abs(pp), levels), title(s, 'FontSize', fs), colormap([0 0 0]);
end

File: p10.py
"""
Polynomials and corresponding equipotential curve

@author: Khiem Nguyen
"""

import numpy as np
import matplotlib.pyplot as plt

N = 16
plt.close('all'); plt.figure(figsize=(10, 8))
for i in range(2):
    if i == 0:
        s = 'equispaced points'
        x = -1 + 2*np.linspace(0, 1, N+1)
    elif i == 1:
        s = 'Chebyshev points'
        x = np.cos(np.pi * np.linspace(0, 1, N+1))
    p = np.poly(x)
    # Plot p(x) over [-1, 1]
    xx = np.linspace(-1, 1, 1001)
    pp = np.polyval(p, xx)
    plt.subplot(2, 2, 2*i + 1)
    plt.plot(x, np.zeros(x.shape), 'o', markersize=4)
    plt.plot(xx, pp, 'k-', linewidth=1.2)
    plt.title(s); plt.grid(True)
    
    # Plot equipotential curves
    plt.subplot(2, 2, 2*i + 2)
    plt.plot(np.real(x), np.imag(x), 'o', markersize=4)
    xg = np.linspace(-1.4, 1.4, 201)
    yg = np.linspace(-1.12, 1.12, 201)
    xx, yy = np.meshgrid(xg, yg)
    pp = np.polyval(p, xx + 1j*yy)
    levels = 10.0 ** np.arange(-4, 1)
    plt.contour(xx, yy, np.abs(pp), levels)
    plt.title(s)